src/threading/affinity.cpp

```diff
@@ -0,0 +1,63 @@
+#include <vector>
+
+#include <cstdlib>
+
+#ifdef __linux__
+
+    #ifndef _GNU_SOURCE
+        #define _GNU_SOURCE
+    #endif
+
+    extern "C" {
+        #include <sched.h>
+    }
+
+#endif
+
+namespace nest {
+namespace mc {
+namespace threading {
+
+#ifdef __linux__
+std::vector<int> get_affinity() {
+    cpu_set_t cpu_set_mask;
+
+    auto status = sched_getaffinity(0, sizeof(cpu_set_t), &cpu_set_mask);
+
+    if(status==-1) {
+        return {};
+    }
+
+    auto cpu_count = CPU_COUNT(&cpu_set_mask);
+
+    std::vector<int> cores;
+    for(auto i=0; i<CPU_SETSIZE && cores.size()<cpu_count; ++i) {
+        if(CPU_ISSET(i, &cpu_set_mask)) {
+            cores.push_back(i);
+        }
+    }
+
+    if(cores.size() != cpu_count) {
+        return {};
+    }
+
+    return cores;
+}
+#else
+
+// No support for non-linux systems
+std::vector<int> get_affinity() {
+    return {};
+}
+#endif
+
+unsigned count_available_cores() {
```

Comment:
I'd return zero here, and then throw an error in the code for deciding on the default number. So, if the code works, it's a good default -- but if it doesn't, you don't get a silent failure ala perl, but a loud failure since in fact the code is not guessing your reasonable intent.